visitor.go

```diff
@@ -79,6 +84,16 @@ func (vis *visitor) visit(w io.Writer, v Value) {
 // It returns true if the value is nil, or recursion has occurred, indicating
 // that the value should not be rendered.
 func (vis *visitor) enter(w io.Writer, v Value) bool {
+	if !v.CanNil() {
+		return false
+	}
+
+	// If the value is an empty interface, return false as there is a specific
+	// case for rendering empty interfaces.
+	if v.DynamicType.Kind() == reflect.Interface && v.Value.IsNil() {
```

Comment:
If this `if` is removed, is the output still correct? Could we remove this, and let the `enter()` function handle rendering the nil interfaces, *and* remove `visitInterface()`?

Reply:
We can definitely do that. However we still need [this condition](https://github.com/dogmatiq/dapper/blob/master/interface.go#L20) from `visitor.visitInterface()` to keep the tests as they are.

Reply:
Hrm, I guess since there is still interface-specific logic it belongs in a `visitInterface()` method, but something about having to do this check twice feels wrong, but I haven't come up with a "better" way yet.

Reply:
I *think* now that this has been made more generic, it probably *shouldn't* be responsible for `nil` rendering at all, only for the recursion.  Which means we should let the `nil` rendering fall through to the filters and main switch statement regardless of the type (ie, not *just* interfaces).  Does that make sense?

Reply:
Yes, definitely it makes sense. I'll commit more changes towards that direction soon.

Reply:
Ok, my last commit is an attempt to let only a single purpose for `visitor.enter()` method - recursion detection and recursive marker rendering.

I had to replace `Value.canNil()` with `Value.canPointer()` method since `visitor.enter()` no longer needs to check the value for `nil`, but it still needs to check if we can safely acquire e a pointer from the value.